File: internal/app/highlight.go
package app

import (
	"strings"

	"github.com/charmbracelet/lipgloss"

	"tdb/internal/config"
	"tdb/internal/suggest"
)

// synClass is the lexical class of a byte in the query buffer, used to colorize
// the editor. The keyword/function/command sets come from the vendored syntax
// data (internal/suggest), so highlighting and completion share one source.
type synClass uint8

const (
	synPlain synClass = iota
	synKw
	synFn
	synStr
	synNum
	synCom
	synVar
	synPun
	synKey  // JSON object key ("name":)
	synBool // JSON true / false / null
)

func isIdentByte(b byte) bool {
	return b == '_' || (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9')
}

func isIdentStart(b byte) bool {
	return b == '_' || (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z')
}

// highlightClasses returns a per-byte lexical class for the buffer under the
// given driver. Unknown drivers get all-plain (no highlight).
func highlightClasses(driver config.Driver, text string) []synClass {
	cls := make([]synClass, len(text))
	switch driver {
	case config.DriverMySQL, config.DriverDoris:
		highlightSQL(text, suggest.SQLKeywordSet(driver), suggest.SQLFunctionSet(driver), cls)
	case config.DriverRedis:
		highlightRedis(text, suggest.RedisCommandSet(), cls)
	case config.DriverMongo:
		highlightMongo(text, suggest.MongoMethodSet(), cls)
	}
	return cls
}

func fill(cls []synClass, start, end int, c synClass) {
	for i := start; i < end && i < len(cls); i++ {
		cls[i] = c
	}
}

// scanString returns the index just past a quoted string starting at i (quote =
// text[i]). It tolerates doubled-quote and backslash escapes.
func scanString(text string, i int) int {
	quote := text[i]
	i++
	for i < len(text) {
		if text[i] == '\\' && i+1 < len(text) {
			i += 2
			continue
		}
		if text[i] == quote {
			if i+1 < len(text) && text[i+1] == quote { // doubled quote escape
				i += 2
				continue
			}
			return i + 1
		}
		i++
	}
	return len(text)
}

func scanNumber(text string, i int) int {
	for i < len(text) {
		b := text[i]
		if (b >= '0' && b <= '9') || b == '.' || b == 'x' || b == 'X' || b == 'e' || b == 'E' ||
			((b == '+' || b == '-') && i > 0 && (text[i-1] == 'e' || text[i-1] == 'E')) ||
			(b >= 'a' && b <= 'f') || (b >= 'A' && b <= 'F') {
			i++
			continue
		}
		break
	}
	return i
}

func nextNonSpace(text string, i int) byte {
	for i < len(text) {
		if text[i] != ' ' && text[i] != '\t' && text[i] != '\n' {
			return text[i]
		}
		i++
	}
	return 0
}

const sqlOperatorBytes = "+-*/%=<>!&|^~"
const sqlPunctBytes = "(),;."

func highlightSQL(text string, keywords, functions map[string]bool, cls []synClass) {
	i := 0
	for i < len(text) {
		b := text[i]
		switch {
		case b == '-' && i+1 < len(text) && text[i+1] == '-', b == '#':
			end := strings.IndexByte(text[i:], '\n')
			if end < 0 {
				end = len(text)
			} else {
				end = i + end
			}
			fill(cls, i, end, synCom)
			i = end
		case b == '/' && i+1 < len(text) && text[i+1] == '*':
			end := strings.Index(text[i+2:], "*/")
			if end < 0 {
				end = len(text)
			} else {
				end = i + 2 + end + 2
			}
			fill(cls, i, end, synCom)
			i = end
		case b == '\'' || b == '"':
			end := scanString(text, i)
			fill(cls, i, end, synStr)
			i = end
		case b == '`':
			end := scanString(text, i)
			i = end // backtick identifiers stay plain
		case b >= '0' && b <= '9':
			end := scanNumber(text, i)
			fill(cls, i, end, synNum)
			i = end
		case isIdentStart(b):
			j := i
			for j < len(text) && isIdentByte(text[j]) {
				j++
			}
			word := strings.ToUpper(text[i:j])
			switch {
			case keywords[word]:
				fill(cls, i, j, synKw)
			case functions[word] || nextNonSpace(text, j) == '(':
				fill(cls, i, j, synFn)
			}
			i = j
		case strings.IndexByte(sqlOperatorBytes, b) >= 0:
			fill(cls, i, i+1, synPun)
			i++
		case strings.IndexByte(sqlPunctBytes, b) >= 0:
			fill(cls, i, i+1, synPun)
			i++
		default:
			i++
		}
	}
}

func highlightMongo(text string, methods map[string]bool, cls []synClass) {
	i := 0
	for i < len(text) {
		b := text[i]
		switch {
		case b == '/' && i+1 < len(text) && text[i+1] == '/':
			end := strings.IndexByte(text[i:], '\n')
			if end < 0 {
				end = len(text)
			} else {
				end = i + end
			}
			fill(cls, i, end, synCom)
			i = end
		case b == '/' && i+1 < len(text) && text[i+1] == '*':
			end := strings.Index(text[i+2:], "*/")
			if end < 0 {
				end = len(text)
			} else {
				end = i + 2 + end + 2
			}
			fill(cls, i, end, synCom)
			i = end
		case b == '\'' || b == '"':
			end := scanString(text, i)
			fill(cls, i, end, synStr)
			i = end
		case b == '$':
			j := i + 1
			for j < len(text) && isIdentByte(text[j]) {
				j++
			}
			fill(cls, i, j, synVar)
			i = j
		case b >= '0' && b <= '9':
			end := scanNumber(text, i)
			fill(cls, i, end, synNum)
			i = end
		case isIdentStart(b):
			j := i
			for j < len(text) && isIdentByte(text[j]) {
				j++
			}
			if methods[strings.ToUpper(text[i:j])] {
				fill(cls, i, j, synFn)
			}
			i = j
		case strings.IndexByte("{}[]():,", b) >= 0:
			fill(cls, i, i+1, synPun)
			i++
		default:
			i++
		}
	}
}

func highlightRedis(text string, commands map[string]bool, cls []synClass) {
	i := 0
	first := true
	for i < len(text) {
		b := text[i]
		switch {
		case b == ' ' || b == '\t' || b == '\n':
			i++
		case b == '\'' || b == '"':
			end := scanString(text, i)
			fill(cls, i, end, synStr)
			i = end
			first = false
		case b >= '0' && b <= '9':
			end := scanNumber(text, i)
			fill(cls, i, end, synNum)
			i = end
			first = false
		default:
			j := i
			for j < len(text) && text[j] != ' ' && text[j] != '\t' && text[j] != '\n' {
				j++
			}
			if first && commands[strings.ToUpper(text[i:j])] {
				fill(cls, i, j, synKw)
			}
			first = false
			i = j
		}
	}
}

// synStyle returns the style for a class and whether it should be applied.
func (m *Model) synStyle(theme appTheme, c synClass) (lipgloss.Style, bool) {
	return synStyleFor(theme, c)
}

// synStyleFor is the Model-free style lookup so renderers without a Model (e.g.
// applyHighlight on JSON output) can colorize too.
func synStyleFor(theme appTheme, c synClass) (lipgloss.Style, bool) {
	switch c {
	case synKw:
		return theme.synKeyword, true
	case synFn:
		return theme.synFunction, true
	case synStr:
		return theme.synString, true
	case synNum:
		return theme.synNumber, true
	case synCom:
		return theme.synComment, true
	case synVar:
		return theme.synVariable, true
	case synPun:
		return theme.synPunct, true
	case synKey:
		return theme.synKey, true
	case synBool:
		return theme.synBool, true
	default:
		return lipgloss.Style{}, false
	}
}

// highlightJSON classifies pretty-printed JSON (post wrapMongoObjectIDs) into
// lexical spans: object keys vs string values, numbers, true/false/null, the
// {}[],: punctuation, and mongo shell constructors like ObjectId("…") / ISODate(…)
// (identifier immediately followed by "(") which render as functions.
func highlightJSON(text string, cls []synClass) {
	i := 0
	for i < len(text) {
		b := text[i]
		switch {
		case b == '"':
			end := scanString(text, i)
			if nextNonSpace(text, end) == ':' {
				fill(cls, i, end, synKey)
			} else {
				fill(cls, i, end, synStr)
			}
			i = end
		case b == '-' || (b >= '0' && b <= '9'):
			start := i
			if b == '-' {
				i++
			}
			end := scanNumber(text, i)
			if end <= start {
				end = start + 1
			}
			fill(cls, start, end, synNum)
			i = end
		case isIdentStart(b):
			j := i
			for j < len(text) && isIdentByte(text[j]) {
				j++
			}
			word := text[i:j]
			switch {
			case nextNonSpace(text, j) == '(':
				fill(cls, i, j, synFn) // ObjectId( / ISODate( / NumberLong( …
			case word == "true" || word == "false" || word == "null":
				fill(cls, i, j, synBool)
			}
			i = j
		case strings.IndexByte("{}[]():,", b) >= 0:
			fill(cls, i, i+1, synPun)
			i++
		default:
			i++
		}
	}
}

// jsonClasses tokenizes a whole JSON document into a per-byte class slice.
func jsonClasses(text string) []synClass {
	cls := make([]synClass, len(text))
	highlightJSON(text, cls)
	return cls
}

// applyHighlight composes text with its per-byte classes into an ANSI-styled
// string. Consecutive bytes of the same class are coalesced into one Render call
// — class boundaries always fall on rune boundaries (fill() colors whole tokens),
// so each segment is valid UTF-8 and CJK is emitted whole. Coalescing keeps the
// ANSI volume low for large JSON and avoids per-rune SGR resets. Colored tokens
// never span a newline (JSON escapes them), so split-by-line stays self-contained.
func applyHighlight(text string, classes []synClass, theme appTheme) string {
	var b strings.Builder
	for i := 0; i < len(text); {
		c := classes[i]
		j := i
		for j < len(text) && classes[j] == c {
			j++
		}
		seg := text[i:j]
		if style, ok := synStyleFor(theme, c); ok {
			b.WriteString(style.Render(seg))
		} else {
			b.WriteString(seg)
		}
		i = j
	}
	return b.String()
}

// highlightJSONText is the convenience one-shot: tokenize + colorize JSON.
func highlightJSONText(text string, theme appTheme) string {
	return applyHighlight(text, jsonClasses(text), theme)
}

File: src/app/highlight.rs
//! Editor syntax highlighting. Port of `internal/app/highlight.go`. The keyword
//! / function / command sets come from the vendored `suggest` data so
//! highlighting and completion share one source. Classification is per byte; it
//! never alters the text (contract §15).

use ratatui::style::{Color, Modifier, Style};
use ratatui::text::Span;

use crate::config::Driver;
use crate::suggest;

use super::theme;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SynClass {
    Plain,
    Kw,
    Fn,
    Str,
    Num,
    Com,
    Var,
    Pun,
    Key,
    Bool,
}

fn is_ident_byte(b: u8) -> bool {
    b == b'_' || b.is_ascii_alphanumeric()
}
fn is_ident_start(b: u8) -> bool {
    b == b'_' || b.is_ascii_alphabetic()
}

/// Per-byte lexical class for `text` under `driver` (None = SQL drivers via the
/// shared keyword/function sets). Unknown drivers get all-plain.
pub fn highlight_classes(driver: Option<Driver>, text: &str) -> Vec<SynClass> {
    let mut cls = vec![SynClass::Plain; text.len()];
    match driver {
        Some(Driver::Mysql) | Some(Driver::Doris) => {
            let d = driver.unwrap();
            highlight_sql(text, suggest::sql_keyword_set(d), suggest::sql_function_set(d), &mut cls);
        }
        Some(Driver::Redis) => highlight_redis(text, suggest::redis_command_set(), &mut cls),
        Some(Driver::Mongo) => highlight_mongo(text, suggest::mongo_method_set(), &mut cls),
        None => {}
    }
    cls
}

pub fn highlight_json_classes(text: &str) -> Vec<SynClass> {
    let mut cls = vec![SynClass::Plain; text.len()];
    highlight_json(text, &mut cls);
    cls
}

fn fill(cls: &mut [SynClass], start: usize, end: usize, c: SynClass) {
    let len = cls.len();
    for slot in cls.iter_mut().take(end.min(len)).skip(start) {
        *slot = c;
    }
}

fn scan_string(text: &[u8], mut i: usize) -> usize {
    let quote = text[i];
    i += 1;
    while i < text.len() {
        if text[i] == b'\\' && i + 1 < text.len() {
            i += 2;
            continue;
        }
        if text[i] == quote {
            if i + 1 < text.len() && text[i + 1] == quote {
                i += 2;
                continue;
            }
            return i + 1;
        }
        i += 1;
    }
    text.len()
}

fn scan_number(text: &[u8], mut i: usize) -> usize {
    while i < text.len() {
        let b = text[i];
        let prev_e = i > 0 && (text[i - 1] == b'e' || text[i - 1] == b'E');
        if b.is_ascii_digit()
            || b == b'.'
            || matches!(b, b'x' | b'X' | b'e' | b'E')
            || ((b == b'+' || b == b'-') && prev_e)
            || (b as char).is_ascii_hexdigit()
        {
            i += 1;
        } else {
            break;
        }
    }
    i
}

fn next_non_space(text: &[u8], mut i: usize) -> u8 {
    while i < text.len() {
        if !matches!(text[i], b' ' | b'\t' | b'\n') {
            return text[i];
        }
        i += 1;
    }
    0
}

const SQL_OPERATOR: &[u8] = b"+-*/%=<>!&|^~";
const SQL_PUNCT: &[u8] = b"(),;.";

fn highlight_sql(text: &str, keywords: &std::collections::HashSet<String>, functions: &std::collections::HashSet<String>, cls: &mut [SynClass]) {
    let t = text.as_bytes();
    let mut i = 0;
    while i < t.len() {
        let b = t[i];
        if (b == b'-' && i + 1 < t.len() && t[i + 1] == b'-') || b == b'#' {
            let end = text[i..].find('\n').map(|e| i + e).unwrap_or(t.len());
            fill(cls, i, end, SynClass::Com);
            i = end;
        } else if b == b'/' && i + 1 < t.len() && t[i + 1] == b'*' {
            let end = text[i + 2..].find("*/").map(|e| i + 2 + e + 2).unwrap_or(t.len());
            fill(cls, i, end, SynClass::Com);
            i = end;
        } else if b == b'\'' || b == b'"' {
            let end = scan_string(t, i);
            fill(cls, i, end, SynClass::Str);
            i = end;
        } else if b == b'`' {
            i = scan_string(t, i);
        } else if b.is_ascii_digit() {
            let end = scan_number(t, i);
            fill(cls, i, end, SynClass::Num);
            i = end;
        } else if is_ident_start(b) {
            let mut j = i;
            while j < t.len() && is_ident_byte(t[j]) {
                j += 1;
            }
            let word = text[i..j].to_uppercase();
            if keywords.contains(&word) {
                fill(cls, i, j, SynClass::Kw);
            } else if functions.contains(&word) || next_non_space(t, j) == b'(' {
                fill(cls, i, j, SynClass::Fn);
            }
            i = j;
        } else if SQL_OPERATOR.contains(&b) || SQL_PUNCT.contains(&b) {
            fill(cls, i, i + 1, SynClass::Pun);
            i += 1;
        } else {
            i += 1;
        }
    }
}

fn highlight_mongo(text: &str, methods: &std::collections::HashSet<String>, cls: &mut [SynClass]) {
    let t = text.as_bytes();
    let mut i = 0;
    while i < t.len() {
        let b = t[i];
        if b == b'/' && i + 1 < t.len() && t[i + 1] == b'/' {
            let end = text[i..].find('\n').map(|e| i + e).unwrap_or(t.len());
            fill(cls, i, end, SynClass::Com);
            i = end;
        } else if b == b'/' && i + 1 < t.len() && t[i + 1] == b'*' {
            let end = text[i + 2..].find("*/").map(|e| i + 2 + e + 2).unwrap_or(t.len());
            fill(cls, i, end, SynClass::Com);
            i = end;
        } else if b == b'\'' || b == b'"' {
            let end = scan_string(t, i);
            fill(cls, i, end, SynClass::Str);
            i = end;
        } else if b == b'$' {
            let mut j = i + 1;
            while j < t.len() && is_ident_byte(t[j]) {
                j += 1;
            }
            fill(cls, i, j, SynClass::Var);
            i = j;
        } else if b.is_ascii_digit() {
            let end = scan_number(t, i);
            fill(cls, i, end, SynClass::Num);
            i = end;
        } else if is_ident_start(b) {
            let mut j = i;
            while j < t.len() && is_ident_byte(t[j]) {
                j += 1;
            }
            if methods.contains(&text[i..j].to_uppercase()) {
                fill(cls, i, j, SynClass::Fn);
            }
            i = j;
        } else if b"{}[]():,".contains(&b) {
            fill(cls, i, i + 1, SynClass::Pun);
            i += 1;
        } else {
            i += 1;
        }
    }
}

fn highlight_redis(text: &str, commands: &std::collections::HashSet<String>, cls: &mut [SynClass]) {
    let t = text.as_bytes();
    let mut i = 0;
    let mut first = true;
    while i < t.len() {
        let b = t[i];
        if matches!(b, b' ' | b'\t' | b'\n') {
            i += 1;
        } else if b == b'\'' || b == b'"' {
            let end = scan_string(t, i);
            fill(cls, i, end, SynClass::Str);
            i = end;
            first = false;
        } else if b.is_ascii_digit() {
            let end = scan_number(t, i);
            fill(cls, i, end, SynClass::Num);
            i = end;
            first = false;
        } else {
            let mut j = i;
            while j < t.len() && !matches!(t[j], b' ' | b'\t' | b'\n') {
                j += 1;
            }
            if first && commands.contains(&text[i..j].to_uppercase()) {
                fill(cls, i, j, SynClass::Kw);
            }
            first = false;
            i = j;
        }
    }
}

fn highlight_json(text: &str, cls: &mut [SynClass]) {
    let t = text.as_bytes();
    let mut i = 0;
    while i < t.len() {
        let b = t[i];
        if b == b'"' {
            let end = scan_string(t, i);
            if next_non_space(t, end) == b':' {
                fill(cls, i, end, SynClass::Key);
            } else {
                fill(cls, i, end, SynClass::Str);
            }
            i = end;
        } else if b == b'-' || b.is_ascii_digit() {
            let start = i;
            if b == b'-' {
                i += 1;
            }
            let mut end = scan_number(t, i);
            if end <= start {
                end = start + 1;
            }
            fill(cls, start, end, SynClass::Num);
            i = end;
        } else if is_ident_start(b) {
            let mut j = i;
            while j < t.len() && is_ident_byte(t[j]) {
                j += 1;
            }
            let word = &text[i..j];
            if next_non_space(t, j) == b'(' {
                fill(cls, i, j, SynClass::Fn); // ObjectId(...) / ISODate(...)
            } else if matches!(word, "true" | "false" | "null") {
                fill(cls, i, j, SynClass::Bool);
            }
            i = j;
        } else if b"{}[],:".contains(&b) {
            fill(cls, i, i + 1, SynClass::Pun);
            i += 1;
        } else {
            i += 1;
        }
    }
}

fn style_for(c: SynClass) -> Option<Style> {
    let s = match c {
        SynClass::Kw => Style::default().fg(theme::ACCENT).add_modifier(Modifier::BOLD),
        SynClass::Fn => Style::default().fg(theme::VIOLET),
        SynClass::Str => Style::default().fg(theme::SUCCESS),
        SynClass::Num => Style::default().fg(theme::WARNING),
        SynClass::Com => Style::default().fg(theme::MUTED).add_modifier(Modifier::ITALIC),
        SynClass::Var => Style::default().fg(theme::KEY),
        SynClass::Pun => Style::default().fg(theme::MUTED),
        SynClass::Key => Style::default().fg(theme::INFO),
        SynClass::Bool => Style::default().fg(theme::VIOLET),
        SynClass::Plain => return None,
    };
    Some(s)
}

/// Build styled spans for one line of `text` given its per-byte classes,
/// grouping consecutive same-class bytes and snapping to char boundaries.
pub fn spans_for_line(text: &str, classes: &[SynClass]) -> Vec<Span<'static>> {
    let mut spans = Vec::new();
    let t = text.as_bytes();
    let mut i = 0;
    while i < t.len() {
        let c = classes.get(i).copied().unwrap_or(SynClass::Plain);
        let mut j = i + 1;
        while j < t.len() && classes.get(j).copied().unwrap_or(SynClass::Plain) == c && text.is_char_boundary(j) {
            j += 1;
        }
        // snap to char boundary
        while j < t.len() && !text.is_char_boundary(j) {
            j += 1;
        }
        let seg = text[i..j].to_string();
        match style_for(c) {
            Some(style) => spans.push(Span::styled(seg, style)),
            None => spans.push(Span::styled(seg, Style::default().fg(Color::Reset))),
        }
        i = j;
    }
    spans
}

#[cfg(test)]
mod tests {
    use super::*;

    fn classes(driver: Driver, text: &str) -> Vec<SynClass> {
        highlight_classes(Some(driver), text)
    }

    #[test]
    fn sql_keyword_and_function() {
        let text = "SELECT COUNT(*) FROM t";
        let c = classes(Driver::Mysql, text);
        assert_eq!(c[0], SynClass::Kw); // SELECT
        let count_idx = text.find("COUNT").unwrap();
        assert_eq!(c[count_idx], SynClass::Fn);
        let from_idx = text.find("FROM").unwrap();
        assert_eq!(c[from_idx], SynClass::Kw);
    }

    #[test]
    fn sql_number_and_comment() {
        let text = "select 42 -- note";
        let c = classes(Driver::Mysql, text);
        assert_eq!(c[text.find("42").unwrap()], SynClass::Num);
        assert_eq!(c[text.find("--").unwrap()], SynClass::Com);
    }

    #[test]
    fn mongo_var_and_method() {
        let text = "db.t.find({ $gt: 5 })";
        let c = classes(Driver::Mongo, text);
        assert_eq!(c[text.find("$gt").unwrap()], SynClass::Var);
        assert_eq!(c[text.find("find").unwrap()], SynClass::Fn);
        assert_eq!(c[text.find('5').unwrap()], SynClass::Num);
    }

    #[test]
    fn json_key_string_bool_and_objectid() {
        let text = r#"{"_id": ObjectId("x"), "ok": true, "n": 1}"#;
        let c = highlight_json_classes(text);
        assert_eq!(c[text.find("\"_id\"").unwrap()], SynClass::Key);
        assert_eq!(c[text.find("ObjectId").unwrap()], SynClass::Fn);
        assert_eq!(c[text.find("true").unwrap()], SynClass::Bool);
    }

    #[test]
    fn highlight_preserves_text_length() {
        // CJK must not break classification or span reconstruction (contract §1).
        let text = "select '真人' from t";
        let c = classes(Driver::Mysql, text);
        assert_eq!(c.len(), text.len());
        let spans = spans_for_line(text, &c);
        let rebuilt: String = spans.iter().map(|s| s.content.as_ref()).collect();
        assert_eq!(rebuilt, text);
    }
}
